diff --git a/gtk/gtkfilechoosernative.c b/gtk/gtkfilechoosernative.c
index cfa274f..94c8df2 100644
--- a/gtk/gtkfilechoosernative.c
+++ b/gtk/gtkfilechoosernative.c
@@ -466,6 +466,7 @@ gtk_file_chooser_native_set_property (GObject      *object,
 
     case GTK_FILE_CHOOSER_PROP_FILTER:
       self->current_filter = g_value_get_object (value);
+      gtk_file_chooser_set_filter (GTK_FILE_CHOOSER (self->dialog), self->current_filter);
       g_object_notify (G_OBJECT (self), "filter");
       break;
 
@@ -494,6 +495,7 @@ gtk_file_chooser_native_get_property (GObject    *object,
       break;
 
     case GTK_FILE_CHOOSER_PROP_FILTER:
+      self->current_filter = gtk_file_chooser_get_filter (GTK_FILE_CHOOSER (self->dialog));
       g_value_set_object (value, self->current_filter);
       break;
 